File: export_MEGA_format.R
# 清空工作空间
rm(list=ls())
# 安装R包
install.packages("seqinr")
# 加载序列数据
S <- seqinr::read.fasta(file.choose())
# 提取特征向量
get_tz <- function(m){
  lst <- list()
  for (i in 1:m) {
    n <- numeric(6)
    weizhi <- numeric(6)
    for (j in 1:length(S[[i]])) {
      s <- S[[i]][j]
      if (s %in% c("a","i","l","m","v","f")){
        n[1]=n[1]+1
        weizhi[1]=weizhi[1]+j
      }
      if (s %in% c("y")){
        n[2]=n[2]+1
        weizhi[2]=weizhi[2]+j
      }
      if (s %in% c("q","n","s","g","t","p","w")){
        n[3]=n[3]+1
        weizhi[3]=weizhi[3]+j
      }
      if (s %in% c("k","r","h")){
        n[4]=n[4]+1
        weizhi[4]=weizhi[4]+j
      }
      if (s %in% c("e","d")){
        n[5]=n[5]+1
        weizhi[5]=weizhi[5]+j
      }
      if (s %in% c("c")){
        n[6]=n[6]+1
        weizhi[6]=weizhi[6]+j
      }
    }
    
    lst[[i]] <- n/length(S[[i]])
    lst[[i]][7:12] <- weizhi/n
  }
  return(lst)
}

# 运行函数所需时间
system.time(tz <- get_tz(length(S)))
tz <- get_tz(length(S)) |> data.frame() |> t()

# 第一步：导出抬头格式 
geshi <- c("#mega","!Title:;")
write.table(geshi,"dist.meg",quote = F,row.names = F,col.names = F)

# 第二步：导出序列描述
r_name <- paste0("[",1:35,"]")

description <- seqinr::getAnnot(S)
description <- gsub(">","#",description) # 将>批量转换为#
description <- gsub(" ","_",description) # 将空格批量转换为_
description <- data.frame(description)
row.names(description) <- r_name
write.table(description,"dist.meg",append = T,quote = F,col.names = F)

# 第三步：导出矩阵的列抬头
c_name <- c("[",1:35,"]")
write.table(t(c_name),"dist.meg",append = T,quote = F,row.names = F,col.names = F,sep = "\t")

row.names(tz) <- r_name # 将行名转为[]
# 用欧氏距离定义样本间的距离
dt <- dist(scale(tz))
dt <- data.matrix(dt)
dt[upper.tri(dt,diag = T)] <- NA # 上三角矩阵和主对角线都赋值为NA

# 第四步：导出矩阵
write.table(round(dt,4),"dist.meg",append = T,quote = F,col.names = F,na="",sep = "\t")
# 将矩阵写入到文件中，让NA值显示为空，间隔为制表符

File: export_MEGA_format.py
import numpy as np
import pandas as pd
from scipy.spatial.distance import pdist, squareform
from Bio import SeqIO
import time
import os

os.chdir("D:\hierarchical-clustering")
# 加载序列数据
sequences = list(SeqIO.parse("influenza_A_virus.fasta", "fasta"))

# 提取特征向量
def get_tz(sequences):
    lst = []
    for seq in sequences:
        n = np.zeros(6)
        weizhi = np.zeros(6)
        for i, s in enumerate(seq.seq):
            if s in "AILMVF":
                n[0] += 1
                weizhi[0] += i + 1
            if s == "Y":
                n[1] += 1
                weizhi[1] += i + 1
            if s in "QNSGTPW":
                n[2] += 1
                weizhi[2] += i + 1
            if s in "KRH":
                n[3] += 1
                weizhi[3] += i + 1
            if s in "ED":
                n[4] += 1
                weizhi[4] += i + 1
            if s == "C":
                n[5] += 1
                weizhi[5] += i + 1
        feature_vector = np.concatenate([n / len(seq), weizhi / n])
        lst.append(feature_vector)
    return np.array(lst)

# 运行函数所需时间
start_time = time.time()
tz = get_tz(sequences)
end_time = time.time()
print(f"运行时间: {round(end_time - start_time,4)} 秒")

# 转换为数据框
tz_df = pd.DataFrame(tz)

# 第一步：导出抬头格式
with open("dist.meg", "w") as f:
    f.write("#mega\n!Title:;\n")

# 第二步：导出序列描述
description = [f"#{rec.description.replace(' ', '_')}" for rec in sequences]
r_name = [f"[{i+1}]" for i in range(len(sequences))]

description_df = pd.DataFrame(description, index=r_name)
description_df.to_csv("dist.meg", mode="a", sep="\t", header=False, index=True)

# 第三步：导出矩阵的列抬头
c_name = '\t'.join(str(i+1) for i in range(len(sequences)))
with open("dist.meg", "a") as f:
    f.write(f"[{c_name}]\n")

tz_df.index = r_name

# 用欧氏距离定义样本间的距离
scaled_tz = (tz_df - tz_df.mean()) / tz_df.std()  # 标准化数据
dt = pdist(scaled_tz, metric='euclidean')
dt_matrix = squareform(dt)
# 将上三角矩阵和主对角线都赋值为NA
tri_upper_indices = np.triu_indices_from(dt_matrix)
dt_matrix[tri_upper_indices] = np.nan

# 第四步：导出矩阵
dt_df = pd.DataFrame(dt_matrix, index=r_name, columns=r_name)
dt_df.to_csv("dist.meg", mode="a", sep="\t", header=False, index=True, float_format="%.4f", na_rep="")
